issue/3812-variation-detail-from-order

nbradbury opened the pull request:
Resolves #3812 by using deep linking to navigate to the variation detail when a variation is clicked in the order detail.

To test:

- Open order detail for an order that contains a variation
- Tap on the product variation
- Verify the variation detail appears and works correctly

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury committed:
Pass productId rather than variationId when opening a product from an order

nbradbury committed:
Set the start destinion for the product navgraph so we go to variation detail

nbradbury committed:
First pass at using a deep link

nbradbury committed:
Navigating deeplink to variation detail now works

nbradbury committed:
added comment

nbradbury committed:
Removed unused import

nbradbury committed:
Removed unused action_global_productVariationDetailFragment from main navgraph

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/3812-variation-detail-from-order

nbradbury committed:
Shorten the deep link host

nbradbury committed:
Updated release notes

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:
```diff
@@ -65,7 +65,10 @@ class VariationDetailViewModel @AssistedInject constructor(
+    private var originalVariation: ProductVariation = variationRepository.getVariation(
+        navArgs.remoteProductId,
+        navArgs.remoteVariationId
+    )!!
```
Not sure if I like using **!!** here.

hichamboushaba replied:
@nbradbury can this be the cause of the crash Ondrej and you noticed?
If we try to navigate to a variation that hasn't been cached before, we will have a NullPointerException, I think the screen needs a loading state where it can fetch the variation if it wasn't cached before.

nbradbury replied:
Yes, this is the cause and I started on a fix a little while ago.

nbradbury committed:
Fixed default arguments in VariationDetailViewModelTest

nbradbury committed:
Mock VariationDetailRepository

nbradbury committed:
Removed unused import

nbradbury committed:
Corrected release notes

0nko reviewed (commented):
After trying to open a variation from an order the app crashed with the following stacktrace:

```
    Process: com.woocommerce.android.prealpha, PID: 13010
    java.lang.NullPointerException
        at com.woocommerce.android.ui.products.variations.VariationDetailViewModel.<init>(VariationDetailViewModel.kt:68)
        at com.woocommerce.android.ui.products.variations.VariationDetailViewModel_Factory.newInstance(VariationDetailViewModel_Factory.java:72)
        at com.woocommerce.android.ui.products.variations.VariationDetailViewModel_Factory.get(VariationDetailViewModel_Factory.java:53)
        at com.woocommerce.android.ui.products.variations.VariationDetailViewModel_Factory_Impl.create(VariationDetailViewModel_Factory_Impl.java:25)
        at com.woocommerce.android.ui.products.variations.VariationDetailViewModel_Factory_Impl.create(VariationDetailViewModel_Factory_Impl.java:8)
        at com.woocommerce.android.viewmodel.ViewModelFactory.create(ViewModelFactory.kt:26)
        at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:69)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:185)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:54)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41)
        at com.woocommerce.android.ui.products.variations.VariationDetailFragment.getViewModel(Unknown Source:2)
        at com.woocommerce.android.ui.products.variations.VariationDetailFragment.initializeViewModel(VariationDetailFragment.kt:133)
        at com.woocommerce.android.ui.products.variations.VariationDetailFragment.onViewCreated(VariationDetailFragment.kt:83)
        at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:2987)
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:546)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:282)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2106)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

I'm not sure if there's anything special about it, it was just the first one I tried.

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/3812-variation-detail-from-order

nbradbury commented:
Strange. I did my testing on API 28 and this works fine there, but it's crashing on API 21, 23, 29, & 30.

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/3812-variation-detail-from-order

nbradbury committed:
Moved opening product variation detail to its own function

nbradbury committed:
Added getOrFetchVariation

nbradbury committed:
retrieve original variation when null

nbradbury committed:
Enable null variation

nbradbury committed:
Ignore setter when value is null

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/3812-variation-detail-from-order

nbradbury commented:
@hichamboushaba I see you've assigned yourself to this PR, which is fine, but I'd also like to ask @ThomazFB to give it a look since he coded the variation detail screen.

jkmassel commented:
This PR is being bumped to `6.6` because `6.5` is being frozen today.

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/adapter/OrderDetailProductListAdapter.kt:
```diff
@@ -29,12 +28,15 @@ class OrderDetailProductListAdapter(
         holder.view.initView(orderItems[position], productImage, formatCurrencyForDisplay)
         holder.view.setOnClickListener {
-            productItemListener.openOrderProductDetail(productId)
+            if (item.variationId == 0L) {
+                productItemListener.openOrderProductDetail(item.productId)
+            } else {
+                productItemListener.openOrderProductVariationDetail(item.productId, item.variationId)
+            }
```
np: We could use the `isVariation` property from `Order.Item` to improve a little bit the readability for this conditional since it's not directly implied that `0L` means a non-variation product. WDYT?
```suggestion
           if (item.isVariation) {
                productItemListener.openOrderProductVariationDetail(item.productId, item.variationId)
            } else {
                productItemListener.openOrderProductDetail(item.productId)
            }
```

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:
```diff
@@ -256,21 +273,23 @@ class VariationDetailViewModel @AssistedInject constructor(
-            showVariation(originalVariation)
+
+            // show the variation if we were able to get it, otherwise exit
+            if (originalVariation != null) {
+                showVariation(originalVariation!!)
+            } else {
+                triggerEvent(ShowSnackbar(string.variation_detail_fetch_variation_error))
+                triggerEvent(Exit)
+            }
```
This may be far-streched from my part, but since `originalVariation` is a nullable mutable property from `VariationDetailViewModel` we may assume the risk of throwing a `NullPointerException` with this double-bang expression directly to the property itself, taking into account a scenario where suddenly something goes south with the property becoming null in a race condition situation with all possible possibly change points it has. I would recommend using standard function calls to make sure that once we capture the value, we copy it into a closed context that will safely deliver the object to the `showVariation`. Let me know what you think about this.

```suggestion
            originalVariation.let {
                if(it != null) { 
                    showVariation(it) 
                } else {
                    triggerEvent(ShowSnackbar(string.variation_detail_fetch_variation_error))
                    triggerEvent(Exit)
                }
            }
```

There are different ways to do that, but they may be a little more cryptic with the readability:

```suggestion
            originalVariation?.let {
                showVariation(it)
            } ?: run {
                triggerEvent(ShowSnackbar(string.variation_detail_fetch_variation_error))
                triggerEvent(Exit)
            }
```

ThomazFB reviewed (commented):
Hey @nbradbury, I've tested the improvement, and looks good! Thanks for this solution! I left just two code suggestions, let me know what you think about them, thanks!

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/3812-variation-detail-from-order

Conflicts:
	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailRepository.kt

nbradbury committed:
Resolved conflicts and fixed missing null handlers

nbradbury committed:
Use item.isVariation rather than checking variationId

nbradbury committed:
Avoid `!!` when accessing original variation

nbradbury commented:
> np: We could use the `isVariation` property from `Order.Item` to improve a little bit the readability for this conditional since it's not directly implied that `0L` means a non-variation product. WDYT?

@ThomazFB Agreed! Changed in 29840dc.

> This may be far-streched from my part, but since originalVariation is a nullable mutable property from VariationDetailViewModel we may assume the risk of throwing a NullPointerException with this double-bang expression directly to the property itself

A fair point. I implemented your suggestion in 5154376.

I've also resolved the conflicts with `develop` so this is ready for another round.

ThomazFB reviewed (approved):
Hey @nbradbury, thanks for implementing the suggestions! LGTM! since I'm not sure if @hichamboushaba wants to do any further review, I'll leave the PR approved and not merge it right away.

hichamboushaba reviewed (approved):
I reviewed the code, and looks good 👍 
I didn't test the feature, as @ThomazFB already did.
